### PR TITLE
Add initial support for MQTTv5

### PR DESCRIPTION
**Pull request by bobveringa**
This PR adds support for MQTTv5. The implementation is still pretty basic, unoptimized and unfinished, but the general concepts are already in place.

For now, this should not be merged, and this PR mainly serves as a place to further discuss the implementation-specific details.

### CODE REVIEW

**Comment by ebolisa** [comment]
Why esp32 and Rpi pico w dev boards are not included? It shouldn't be too difficult to add them.

**Comment by bobveringa** [comment]
> Why esp32 and Rpi pico w dev boards are not included? It shouldn't be too difficult to add them. 

I don't understand what you mean by this. There is no change to the platforms this library supports.

Could you clarify your question?

**Comment by peterhinch** [comment]
My approach automates this rather than have the user comment out the import. I suggest the following:
```python
class MQTT_base:
    REPUB_COUNT = 0  # TEST
    DEBUG = False

    def __init__(self, config):
        global encode_properties, decode_properties
# code omitted
        self.mqttv5 = config.get("mqttv5")
        self.mqttv5_con_props = config.get("mqttv5_con_props")
        self.topic_alias_maximum = 0
        if self.mqttv5:
            from .mqtt_v5_properties import encode_properties, decode_properties
```
I have tested the concept (albeit that my hack used different naming).

This enables existing code to run with the RAM saving.

**Comment by bobveringa-sf** [comment]
@peterhinch The documentation has been updated. Unless you want to make any additional changes, I would say this good to merge. 

I am unable to mark this PR as ready for review, this is why I have not updated it.

**Comment by bobveringa** [comment]
@peterhinch This is now ready for final review, the documentation update is included.

File: mqtt_as/README.md [hunk 1]
```diff
@@ -691,6 +695,125 @@ async def messages(client):
...
+include properties in published messages, while using MQTTv3.1.1, the properties
+will be ignored.
+
+### 3.6.2 MQTTv5 Properties
+Properties are a new and important feature of MQTTv5. They are used to provide
+additional information about the message, and allow for more advanced features
+such as message expiry, user properties, and response information.
+
+Incoming properties are formatted as a dictionary using the property  identifier
+as the key. The property identifier is an integer that is defined in the MQTTv5
```
Comment (peterhinch) [inline mqtt_as/README.md:757]: There seems to be a missing word here. Should this read "in the MQTTv5 specification"?

Comment (bobveringa) [reply mqtt_as/README.md:757]: Yes, you are correct. It is fixed now.

**Review by peterhinch — APPROVED** [review]

**Pull request merged**